projects/wpdx/src/app/rehab-prio/adm-popup/adm-popup.component.ts

```diff
@@ -71,80 +71,53 @@ export class AdmPopupComponent implements OnChanges {
-              clean_adm2='${this.fq(value.NAME_2)}' group by ${attribute}`);
-    }
-    if (value.NAME_3) {
-      queries.push(`${baseQuery}
-        where clean_country_name='${this.fq(value.NAME_0)}' and clean_adm1='${this.fq(value.NAME_1)}' and 
-              clean_adm2='${this.fq(value.NAME_2)}' and clean_adm3='${this.fq(value.NAME_3)}' group by ${attribute}`);
-    }
-
-    if (value.NAME_4) {
-      queries.push(`${baseQuery}
-        where clean_country_name='${this.fq(value.NAME_0)}' and clean_adm1='${this.fq(value.NAME_1)}' and 
-              clean_adm2='${this.fq(value.NAME_2)}' and clean_adm3='${this.fq(value.NAME_3)}' and clean_adm4='${this.fq(value.NAME_4)}' group by ${attribute}`);
+    for (let i = 0; i < values.length; i++){
+      baseQuery = this.addQuery(values[i], baseQuery, i+1, queries, attribute)
     }
     return queries;
 
   }
 
-  getManagementData(){
-    forkJoin(this.getDataQuery(this.popupProperties, 'management_clean').map(q => this.db.query(q))).subscribe(results => {
-      this.management_data = [];
+  getData(column_name){
+    forkJoin(this.getDataQuery(this.popupProperties, column_name).map(q => this.db.query(q))).subscribe(results => {
+      const data_array = [];
       for (let i = 0; i < results.length; i++){
         const x = [];
-        for (let j = 0; j < results[i].rows.length; j++){
+        for (let j = 0; j < results[i].rows.length; j ++){
           if (results[i].rows[j].count > 0){
-            x.push({'name': results[i].rows[j].management_clean, 'value': results[i].rows[j].count});
-          }
-      }
-        this.management_data.push(x);
-      }
-    })
-  }
-
-  getTechData(){
-    forkJoin(this.getDataQuery(this.popupProperties, 'water_tech_clean').map(q => this.db.query(q))).subscribe(results => {
-      this.tech_data = [];
-      for (let i = 0; i < results.length; i++){
-        const x = [];
-        for (let j = 0; j < results[i].rows.length; j++){
-          if (results[i].rows[j].count > 0){
-            x.push({'name': results[i].rows[j].water_tech_clean, 'value': results[i].rows[j].count});
+            x.push({'name': Object.values(results[i].rows[j])[1], 'value': results[i].rows[j].count});
+            
           }
         }
-        this.tech_data.push(x);
+        data_array.push(x);
       }
-    })
-  }
-
-  getSourceData(){
-    forkJoin(this.getDataQuery(this.popupProperties, 'water_source_clean').map(q => this.db.query(q))).subscribe(results => {
-      this.source_data = [];
-      for (let i = 0; i < results.length; i++){
-        const x = [];
-        for (let j = 0; j < results[i].rows.length; j++){
-          if (results[i].rows[j].count > 0){
-            x.push({'name': results[i].rows[j].water_source_clean, 'value': results[i].rows[j].count});
-          }
-        }
-        this.source_data.push(x);
+      if (column_name === 'water_tech_clean'){
```

Comment:
This is a big ugly :) if this is a generic function, it shouldn't have this specific code.
You can return an `Observable` here and handle the specific handling of the result in `ngOnChanges`.

This means that you don't subscribe here to the `forkJoin`, but only pipe it through a `map` operator.
In `ngOnChanges` you subscribe to the observable and put the results in their correct place.